class-tgm-plugin-activation.php

```diff
@@ -1587,176 +1607,113 @@ public function process_bulk_actions() {
 			if ( 'tgmpa-bulk-install' === $this->current_action() ) {
 				check_admin_referer( 'bulk-' . $this->_args['plural'] );
 
-				// Prep variables to be populated.
 				$plugins_to_install = array();
-				$plugin_installs    = array();
-				$plugin_path        = array();
-				$plugin_name        = array();
 
-				// Look first to see if information has been passed via WP_Filesystem.
-				if ( isset( $_GET['plugins'] ) ) {
-					$plugins = explode( ',', stripslashes( $_GET['plugins'] ) );
-				}
-				// Looks like the user can use the direct method, take from $_POST.
-				elseif ( isset( $_POST['plugin'] ) ) {
-					$plugins = (array) $_POST['plugin'];
-				}
-				// Nothing has been submitted.
-				else {
-					$plugins = array();
-				}
+				if ( isset( $_POST['plugin'] ) && ! empty( $_POST['plugin'] ) ) {
 
-				// Grab information from $_POST if available.
-				if ( isset( $_POST['plugin'] ) ) {
-					foreach ( $plugins as $plugin_data ) {
-						$plugins_to_install[] = explode( ',', $plugin_data );
-					}
+					if ( is_array( $_POST['plugin'] ) ) {
+						$plugins_to_install = (array) $_POST['plugin'];
 
-					foreach ( $plugins_to_install as $plugin_data ) {
-						$plugin_installs[] = $plugin_data[0];
-						$plugin_path[]     = $plugin_data[1];
-						$plugin_name[]     = $plugin_data[2];
+					} elseif ( is_string( $_POST['plugin'] ) ) {
+						// Received via Filesystem page - unflatten array (WP bug #19643)
+						$plugins_to_install = explode( ',', $_POST['plugin'] );
 					}
-				}
-				// Information has been passed via $_GET.
-				else {
-					foreach ( $plugins as $key => $value ) {
-						// Grab plugin slug for each plugin.
-						if ( 0 === ( $key % 3 ) || 0 === $key ) {
-							$plugins_to_install[] = $value;
-							$plugin_installs[]    = $value;
+					
+					// Sanitize the received input
+					$plugins_to_install = array_map( 'sanitize_title', $plugins_to_install );
+
+					// Validate the received input
+					foreach ( $plugins_to_install as $key => $slug ) {
+						// Check if the plugin was registered with TGMPA and remove if not
+						if ( ! isset( $this->tgmpa->plugins[ $slug ] ) ) {
```

Comment:
Logic is correct. We only want to deal with plugins which were registered in TGMPA, if a plugin was not registered by a theme or plugin using TGMPA and still somehow managed to get into the $_POST array, we ignore it as it's probably some sort of hack.